procedures: Configuring workspace target namespace, Provisioning namespaces (#2301)
codenvy/che-docs: modules/administration-guide/pages/configuring-workspace-target-namespace.adoc

:_content-type: CONCEPT
:navtitle: Configuring workspace target {orch-namespace}
:keywords: administration guide, configuring, namespace
:page-aliases: installation-guide:configuring-namespace-strategies, configuring-namespace-strategies

[id="configuring-workspace-target-namespace_{context}"]
= Configuring workspace target {orch-namespace}

For each user, {prod-short} isolates the workspaces in a {orch-namespace}.
You can configure the {orch-namespace} name template.

{prod-short} server checks usernames and IDs for compatibility with {orch-name} objects naming convention before creating a {orch-namespace} from a template:

> *  Incompatible usernames or IDs are reduced to the nearest valid name by replacing groups of unsuitable symbols with the `-` symbol. 

Reviewer: ```suggestion
* Incompatible usernames or IDs are reduced to the nearest valid name by replacing groups of unsuitable symbols with the `-` symbol. 
```

> * The addition of a random 6-symbol suffix prevents IDs from collisions. The result is stored in preferences for reuse.

Reviewer: id are openshift UIDs AFAIK, I do not think we need any suffix here

Author: I observed that Minikube adds the suffix. Therore, valid on Kubernetes too.

* {kubernetes} limits the length of a {orch-namespace} name to 63 characters (this includes the evaluated placeholders). Additionally, the names (after placeholder evaluation) must be valid DNS names.
* OpenShift limits the length further to 49 characters.

> * The `_<userid>_` placeholder is evaluated into a 36 character long UUID string.

Reviewer: Interestingly so we do really support the 36 chars postfix in the namespace names? was it verified?

Reviewer: in any case, as a user / admin, I could not imagine having fun with namespace setup like `dev-26c52e9d-780d-42d2-9481-869357844e53` 
However, I might be missing smth. and in some cases that could potentially be useful

Author: Che on Minikube: `<userid>-aka-<username>-namespace-for-che` translating to

`cgezegvsb2nhba-aka-user3-namespace-for-che-6m2w2b`
 so more a 14 character string.

```
| `workspaceNamespaceDefault` template value | resulting {orch-namespace} example

| <userid>-aka-<username>-namespace | cgezegvsb2nhba-aka-user3-namespace-for-che-6m2w2b

| <username>-namespace | user4-namespace 
```

.Procedure

* Set the `server.workspaceNamespaceDefault` property in CheCluster Custom Resource:
+
[subs="+quotes,+attributes"]
----
...
spec:
  server:
    workspaceNamespaceDefault: __{prod-workspace}__-__<username>__
...

> ----

Reviewer: ~~I do not think we support this. My understanding is that now we only have one namespace per-user strategy~~

Reviewer: oh, sorry it is probably the same as 
```
server:
      customCheProperties:
        CHE_INFRA_KUBERNETES_NAMESPACE_DEFAULT: <username>-dev
```
which is supported

Author: OK

:_content-type: CONCEPT
:navtitle: Configuring workspace target {orch-namespace}
:keywords: administration guide, configuring, namespace
:page-aliases: installation-guide:configuring-namespace-strategies, configuring-namespace-strategies

[id="configuring-workspace-target-namespace_{context}"]
= Configuring workspace target {orch-namespace}

For each user, {prod-short} isolates the workspaces in a {orch-namespace}.

> You can configure the {orch-namespace} name template.

Reviewer 2: Here we should say that the target {orch-namespace} can be pre-created (link to the doc) or {prod-short} can create it the first time the user starts a workspace and the {orch-namespace} name template can be configured by the admin (this doc).

:_content-type: PROCEDURE
:navtitle: Configuring {orch-namespace} name
:keywords: administration guide, configuring, namespace
:page-aliases: 

> [id="configuring-user-{orch-namespace}-name-for-automatic-provisioning_{context}"]

Reviewer 3: ```suggestion
[id="configuring-a-user-{orch-namespace}-name-for-automatic-provisioning_{context}"]
```

> = Configuring user {orch-namespace} name for automatic provisioning

Reviewer 3: ```suggestion
= Configuring a user {orch-namespace} name for automatic provisioning
```
Need to use the indefinite article here because you're providing instructions for a specific example.

> You can configure the {orch-namespace} name template that {prod-short} is using to create the required {orch-namespace} when starting a workspace.

Reviewer 3: ```suggestion
You can configure the {orch-namespace} name template that {prod-short} uses to create the required {orch-namespace} when starting a workspace.
```
- Continuous tense incorrectly implies that the use of the template is a continuous, ongoing process that doesn't end.
- Should `starting a workspace` be `creating a workspace`?

> A valid {orch-namespace} name template follows this convention:

Reviewer 3: ```suggestion
A valid {orch-namespace} name template follows these conventions:
```
- Plural needed because there are multiple conventions in the list
- Also, consider making the items in the list more parallel in structure. Maybe the problem is with the word "conventions".  Not all of the items in the list are conventions. Some are limitations, and some are descriptions of the system processes. Maybe `guidelines` would be more accurate? Or maybe split it into two lists -- one with the naming conventions, and one explaining what the system does with the names to ensure adherence with the conventions?

* The `<username>` or `<userid>` placeholder is mandatory.

> * {prod-short} reduces usernames or IDs incompatible with {orch-name} objects naming convention to the nearest valid name by replacing groups of unsuitable symbols with the `-` symbol. 

Reviewer 3: ```suggestion
* Usernames and IDs cannot contain invalid characters. If the formatting of a username or ID is incompatible with the naming conventions for {orch-name} objects, {prod-short} changes the username or ID to a valid name by replacing incompatible characters with the `-` symbol. 
```
- Consider revising the sentence as suggested. The current sentence is awkwardly constructed with some unclear antecedents and is too "noun heavy" at times. 
- Also consider replacing "reduce" with "change" (the change may not be a reduction but merely a replacement) and "unsuitable" with "incompatible", which in this case is more accurate.

> * {prod-short} evaluates the `<userid>` placeholder into a 14 character long string, and adds a random 6 character long suffix to prevents IDs from collisions. The result is stored in the user preferences for reuse.

Reviewer 3: ```suggestion
* {prod-short} evaluates the `<userid>` placeholder into a 14 character long string, and adds a random six character long suffix to prevent IDs from colliding. The result is stored in the user preferences for reuse.
```
- Per ISG: `In text, use words for zero through nine. In text, use numerals for 10 and greater.`

* {kubernetes} limits the length of a {orch-namespace} name to 63 characters. 

* OpenShift limits the length further to 49 characters.

> * The resulting name is a valid DNS name.

Reviewer 3: ```suggestion
```

This isn't really a convention as suggested by the intro paragraph for the list. It's the result of following the conventions. You can safely omit it because you already said that the list would define the naming conventions.
